### DISCUSSION TITLE
Warp 1.4.0 Release

### DISCUSSION THREAD

**Discussion by mmacklin** [NVIDIA/warp#322] [Announcements]
# Warp 1.4.0 Release
 
We are pleased to announce that the Warp 1.4.0 release is now available on Github and PyPi! This release includes many improvements to our Python language support, as well as significant external contributions from AutoDesk and Baidu. 
 
* **PyPi**: `pip install warp-lang`
* **Github**: https://github.com/nvidia/warp
* **Docs**: https://nvidia.github.io/warp/

## What’s New
 
* **Improved Support for Runtime Code Generation**
  - Runtime code generation is a powerful feature that allows users to do late-stage specialization of their kernels to specific datatypes, dimensions, and functions, often with significant performance benefits. In this release we have significantly expanded support for dynamic creation of kernels, functions, and structures at runtime. For example, users can now directly return Warp kernels from closure functions as follows:

    ```python
    def create_kernel_with_function(f):
        @wp.kernel
        def k(a: wp.array(dtype=float)):
            tid = wp.tid()
            a[tid] = f(a[tid])
        return k

        @wp.func
        def square(x: float):
            return x * x

        @wp.func
        def cube(x: float):
            return x * x * x

        k1 = create_kernel_with_function(square)
        k2 = create_kernel_with_function(cube)
    ```
  - Closure created kernels will automatically capture any constants, functions, or types passed to the kernel.
  - **Documentation**: https://nvidia.github.io/warp/codegen.html#dynamic-kernel-creation

* **Support for Compile Time Expression Evaluation**
  - In addition to improved runtime code generation, Warp 1.4.0 now supports evaluating Python expressions at compile time using the new wp.static() method.  Static expressions may include arbitrary Python code, including dictionary lookups, file operations, etc. Branches dependent on the result of static expressions are resolved at compile time and can completely skip code generation as shown below:

    ```python
    import warp as wp

    available_colors = {"red", "green", "blue"}

    @wp.kernel
    def my_kernel():
        if wp.static("red" in available_colors):
            print("red is available")
        else:
            print("red is not available")
    ```
    
  - Static expressions may return constant values, functions, and datatypes
  - **Documentation**: https://nvidia.github.io/warp/codegen.html#static-expressions

* **Automatic Differentiation Improvements**

  - We continue to expand our support for passing gradients through kernel code. In this release we have added support for auto-differentiation through in-place operations on kernel vector and matrix types, as well as support for linear in-place operations on wp.array() types as requested by our DLSS and BioNemo users.
  - **Documentation**: https://nvidia.github.io/warp/modules/differentiability.html

* **Distributed JAX Integration (AutoDesk)**
  - The AutoDesk XLB team have significantly improved upon our JAX integration, adding support for MGPU and sharding of Warp kernels. 
  - **Documentation**: https://nvidia.github.io/warp/modules/interoperability.html#jax

* **PaddlePaddle Integration (Baidu)**
  - Contributions to add support for the Paddle DL framework, users can now use wp.from_paddle(), wp.to_paddle() to zero-copy pass data between frameworks
  - **Documentation**: https://nvidia.github.io/warp/codegen.html#static-expressions

## Simulation Changes

* **VBD Integrator**
  - Warp 1.4.0 provides a new VBD integrator which is a drop-in replacement for XPBD for implicit optimization-based time-integration. 
  - **Documentation**: https://nvidia.github.io/warp/modules/sim.html#warp.sim.VBDIntegrator

* **Adaptive Sparse Volumes**
  - warp.fem now supports adaptive NanoVDB fields for improved memory efficiency on large problems.
  - **Documentation**: https://nvidia.github.io/warp/modules/fem.html#adaptivity

## What’s Fixed
 
We have made over 32 bug fixes since the 1.3.0 release, including addressing many Github  issue reports:
 
* [GH-284] Support for fp64 atomic_add, atomic_max, and atomic_min.
* [GH-282] Support wp.mod() for vector types.
* [GH-308] Support shadowing builtin functions.
* [GH-310] Allow passing custom launch dimensions to jax_kernel().
* [GH-302] Fix for wp.func erroring out when defining a Tuple as a return type hint.
* [GH-306] Handle --num_tiles 1 in example_render_opengl.py.
* [GH-312] Fix wp.Volume.load_from_numpy() behavior when bg_value is a sequence of values.
* [GH-281] Fix accuracy of 3x3 SVD wp.svd3 with fp64 numbers.
* [GH-287] Fix module hashing when a kernel argument contained a struct array.
* [GH-288] Fix a bug in wp.bvh_query_ray() where the direction instead of the reciprocal direction was used.
* [GH-289] Fix handling of integer indices in the input_output_mask argument to autograd.jacobian and autograd.jacobian_fd.
 
Full [CHANGELOG.md](https://github.com/NVIDIA/warp/blob/main/CHANGELOG.md)

We hope you enjoy this release, and look forward to seeing what you make with Warp. Please send us any feedback and bug reports on our [Issues ](https://github.com/NVIDIA/warp/issues)page.

**Comment by windwhiterain** [comment]
wow, maybe this will go beyond taichi-lang in the future
